### PR TITLE
Use an sqlite database for the media cache

### LINKED ISSUE #10598: Automatic cleaning of media cache
Currently, the media cache is not automatically cleaned (I have folders in my media folder that were created in 2014!). However, with the not-so-recent introduction of translation files as media, and probably the future introduction of server-sent lua files, a lot of media files will get useless relatively fast.

Therefore, I propose to introduce an automatic cleaning of the media cache, which would remove old files. There needs to be some thinking about the heuristics used, because we don't want to evict some useful media when connecting to servers with lots of custom media. If there are time-based heuristics, the time scale used should probably depend on media type: code and translations will probably have far shorter lifespans than textures or models.

### PR DESCRIPTION
**Pull request by Desour**
- Replaces the `cache_path/media/` folder with an sqlite database `cache_path/media_db.sqlite`.
  Most notable new features are cache cleaning and the possibility to store the files in the cache zlib-compressed.
- Mostly fixes #10598. This is just a very simple approach though. Also, by default nothing is deleted yet.
- How the PR works:
  * Instead of `path_cache/media/many_files`, there's now `path_cache/media_db.sqlite`.
  * The db contains together with the file data and sha1 digest also some metadata per media file: a cache-class and a data-format, and also a timestamp (in seconds since epoch) and a expiry counter (see further below):
    - The data-format says how the file data is encoded, ie. raw or zlib-compressed (only raw is used for now).
    - The cache-class (see enum `MediaCacheClass`) determines how long the file persists in the storage.
      * There are two cache-classes for now: `Legacy` (what we currently use) and `Unknown` (should be used if the server suggested a cache-class that we don't know).
      * Cache-class expiry times can be set by the user via settings (can also be "never" to never expire files).
  * The db is accessed twice on startup:
    - When the media announcement is received, all files are looked up via their sha1 digest, and their timestamp is updated.
    - When new media files are received, they are stored in ram. After all files were received, the db is updated with the new files, and some maintenance is done (see next point). This is done in a separate thread, so the game can start before it's done. (Note that updating all files at once is much faster than doing one transaction per file.)
  * The maintenance mentioned above includes cache expiry and vacuuming:
    - After a cache-class dependent time, a file expires (expire counter incremented, timestamp reset).
    - A file is removed after it expired twice.
    - After 1000 deleted files, the db is vacuumed. (There's a counter in a separate db table for db-wide metadata.) TODO: add a setting for this number
- roadmap: count this to "Internal code refactoring"
- ATQTWAA (answers to questions that weren't answered above):
  - Q: What happens with the old media folder?
    A: It just stays there. The db is not populated from the old cache.
  - Q: What happens if two or more minetest instances load media at the same time?
    A: In the worst case, the one that comes later silently goes back to main menu (I hope at least, that there's no worse case). There are some warnings and error messages in the terminal and debug.txt about the database being locked.
  I've tested this by adding a sleep of 10 seconds in the part of the code where the sql transaction is being done.
  - Q: Do you want performance advice? (This is also a NWBAQ (never would be asked question).)
    A: Yes, always.
 - Some more notes about performance:
   * In mtg, looking up the media files takes very roughly 50 ms, which is nothing compared to the ca. 2 s connection time, of which ca. 0.5 s are media loading.
   * Cache-cleaning is most of the times very cheap (ie. roughly 50 ms). It can however take long if nearly a whole fatly filled cache is cleaned. Vacuuming also takes *very* long if there are many files (around 20 s IIRC). This is compensated via multi-threading. The player can still observe this though through a higher cpu load and he has to wait if he quits directly after joining the game.
- Things open to decide:
  * It was suggested on irc to use an expiry logic that keeps a maximum cache size and removes oldest files when the size exceeds the max size.
    - Possible upsides:
      * There's a maximum size.
      * There's an upper limit for the size of files deleted from cache at once: The size of the new files (plus the size of the largest deleted file).
    - Possible downsides:
      * Very old media is kept even if it's never used. => cache needs more disk space
      * If max size is too small, or/and the player has much big media, the cache won't work effectively.
  * Does vacuuming really make sense? It's expensive and Idk how much performance would degrade if we didn't do it.
- A walk through the src code changes:
  * Utility fixes and upgrades:
    - `util/basic_macros.h`: Added `DISABLE_CLASS_MOVE`, similar to `DISABLE_CLASS_COPY`.
    - `threading/thread.{h, cpp}`: Fixed `Thread::start()` blocking indefinitely if the thread's function finishes too fast. (Needed because of a new unittest.)
    - `client.h`, partly `client.cpp`: Added thread vector for async jobs. (Currently only used for one thread ever.)
    - `util/thread.h`: Added `LambdaThread` class for making a `Thread` from a lambda.
    - `util/time_parsing.{h, cpp}`: Added helper for parsing time values, ie. "1.5 ms". (Used for parsing setting values.)
    - `util/Optional.h`: Fixed some problems with `Optional`. Movement semantics should now work. It should match `std::optional` from the too new C++ stdlib. (Needed because I think I needed move semantics somewhere.)
    - `unittest/test_utilities.cpp`: Unittests for most of the above.
    - small parts of `database/database-sqlite3.h`: Added conversion functions for blob data.
  * Main changes:
    - `builtin/settingtypes.txt`, `defaultsettings.cpp`: New settings for expiry heuristics and co..
    - partly `client/client.cpp`, `client/clientmedia.{h, cpp}`: Switched to the cache db, new functions for cache cleaning and co.. There is also some benchmarking code, TODO: remove.
    - `database/database-sqlite3.{h, cpp}`: Added `MediaCacheDatabaseSQLite3` class for the media db. This is where most things happen.
  * Files not mentioned above:
    - `{util, client}/CMakeLists.txt`
    - `client/filecache.{h, cpp}`: Deleted because it's now unused.

## To do

This PR is Ready for deciding whether it suits into the roadmap.

TODO:
* Add an interface for the media db, to allow implementation of other backends.
* Add setting for vacuuming.
* Though there is some code to measure speed **that needs to be removed _before merge_** (the code needs to be removed, not the speed).

## How to test

- First move your old media cache folder, just to be sure.
- Here's a small lua script to insert the old files into the db (db must exist already): https://gist.github.com/Desour/98229b8604c2957ebd4d2791e539bc48
- Start the game. You should notice no new big delay.
- Play around with the new settings. You can look at the db via the cli `sqlite3` tool.
- Read the code.
  There are some (at least one) code paths that are never taken, ie. the decompression which is currently unused. Also, this PR is not the easiest to test by using, so reading the code is important.
- Try to break something.
- Run the unittests (only for a small new utility helper).

### CODE REVIEW

**Comment by nerzhul** [comment]
whereas i understand the features you want to achieve, storing blobs in sqlite is totally counter productive. I suggest instead that to have cache metadata inside but not the files

**Comment by Desour** [comment]
>storing blobs in sqlite is totally counter productive

Could you elaborate more on this please? I don't see what is wrong with storing blobs in sqlite.

**Comment by SmallJoker** [comment]
Isn't this a bit over-engineered? Keep the date of the last - say - 10 server joins and update the modification date of media files when they're used.

Afterwards delete files when `(oldest_server_join_date - cache_file_modification_date) > 30 days`. This way, casual players will not lose the cache right away, and the cache is earliest deleted after 30 days (subject to change) when not visiting a certain server.

**Comment by Desour** [comment]
SmallJoker:
The cache-classes are there because #10598 suggested it:
>If there are time-based heuristics, the time scale used should probably depend on media type: code and translations will probably have far shorter lifespans than textures or models.

Removing the settings and using fixed times would of course also simplify it a little bit. But imo the settings don't add that much complexity.
Apart from those two points, I think your suggestion is more complicated, because I'd had to store the server join dates.
I don't dislike the suggestion though.

**Comment by celeron55** [comment]
Has someone benchmarked performance compared to the old filesystem approach?

**Comment by Desour** [comment]
I haven't benchmarked the whole connection process, but at least the initial step in the media downloader. And I did only test in singleplayer with mtg. I also haven't tested with a very full media cache.

At first the media loading was much slower than with the old filesystem approach, because I've done one sql transaction per file (it took about 7 s). So, I've changed that (see 3rd commit), and now the cache loading only takes about 50 ms on my machine in mtg. This is nearly nothing compared to the ca. 0.4 seconds that the client takes for loading the files into irrlicht stuff or whatever. The benchmark code for that is still there.
The storing of new media files ("update()") was done for each file as it's received. This was also noticeably slow with the one transaction per file approach, so I've also fixed that by storing the file data in ram and updating the cache with one transaction when the media downloader is finished. I haven't measured the time here. However, it doesn't seem like the media cache is a bottleneck here.

**Comment by celeron55** [comment]
Well, anyway - the problem with the simple timestamp + expiry time approach for deletion is, like discussed in #10598, that if a user stops playing for however long the cache expiry time is, the client shouldn't clear the cache before the user has actually used the client again on multiple servers so that the client actually knows which files are not needed.

One solution to such irregular usage patterns is a two-stage expiration logic, where reaching the expiration time marks the file as "expiration pending", and resets the timestamp, and once the expiration time is reached again with the pending flag already set, then the file will be dropped from the cache. Basically, a file has to expire twice before it's dropped.

I don't agree with nerzhul about storing blobs in sqlite being counter productive. It goes against traditional database design guidelines, but sqlite and many other modern databases are perfectly capable of doing this.

**Comment by Desour** [comment]
FYI, I've done some more benchmarking:
- I've used a more filled media cache (ca. 350 MiB, 43041 files, data size per file from 34 B to ca. 7.3 MB, avg. filesize 8.2 KB, median file size 514 B). See title comment (How To Test section) for a script to insert your files into the sqlite db.
- I've still just tested in mtg (plus mesecons modpack).
- I've also measured the cache update with new files and the cache cleaning.
- I'm using an ext4 file system.
- The results are not meant to be precise, but just give an idea about the order of magnitude.
- Results:
  - (The client media loading (the `Client::checkAndLoad` stuff) takes about 0.75 seconds, idk why I got 0.4 seconds earlier, maybe I misread.) 
  - The loading from cache is still just about 50 ms, with the filesystem approach and the sqlite db, no matter if I take the sparsely or fatly filled cache. (This is very inaccurate for the filesystem approach, because the measurement there is mixed up with the client media loading, which takes much longer.)
  - The cache update also takes *about* 50 ms if the new files are not yet cached (if all were found, it doesn't do anything). Updating cache and the clean check (without deleting anything) take longer with a fatly filled (update: 34 ms vs. 68 ms, clean check: 1 ms vs. 54 ms). I haven't tested the filesystem approach with a filled up folder (because deleting the ones to be loaded from it would make too much work for me now); with an empty folder it takes 50 ms.
  - TL;DR for the above 2 points: I didn't find a relevant slow-down.
  - However, if I use a filled up db and set the expire time to 0 s (= delete everything), it takes about 10 s to delete. If I just delete everything from the table in the sqlite3 cli (with `delete from files where true;`), it also takes several seconds. So, it might make sense to do the cache cleaning in a separate thread.

>a two-stage expiration logic

Ah, yes. That's a good and simple idea. I'll implement that.

**Comment by Desour** [comment]
Squashed, rebased and updated title comment.
Waiting for roadmap approval/denial.
(My interest in this isn't that high, so I won't do any more useless work for now.)

**Comment by erlehmann** [comment]
I think this is an extremely overengineered solution for the goal of media cache expiration.

By which I mean, it is probably among the worst solutions for a problem that I have seen so far.

Cache expiration and cleanup can easily be done by reading `atime` from the existing cache folder.

**Comment by erlehmann** [comment]
Note that this also breaks all those cool shell tricks that were used so far, like finding out the filetypes in the cache using `ls |xargs -n50 file --mime --brief |sort |uniq -c`.

**Comment by Desour** [comment]
> > Q: What happens with the old media folder?
> > A: It just stays there. The db is not populated from the old cache.
> 
> To be clear: This implies that this PR, if merged, nearly doubles the size of the disk space being used when the cache is filled again with all of the stuff that was in it before. In addition, all the users on metered or slow connections will have the trouble of downloading stuff that has already been cached. This is unacceptable for me!
> 
> **Instead of saving disk space, this PR makes Minetest use more of it!**

The alternative would be to remove the old folder. It might make sense to do that.

> **Instead of avoiding traffic, this PR causes Minetest to download stuff again!**

I don't care about a one-time download. When you update minetest, or start minetest without a cache, you also have to download something. `¯\_(ツ)_/¯`

> > Q: What happens if two or more minetest instances load media at the same time?
> > A: In the worst case, the one that comes later silently goes back to main menu (I hope at least, that there's no worse case). There are some warnings and error messages in the terminal and debug.txt about the database being locked.
> 
> Game and mod developers need to be able to start several Minetest instances locally, they often rely on it.
> 
> Runnning several Minetest instannces on the same computer so far only required disabling local map saving.
> 
> I interpret the above as: If this PR gets accepted, it will make local multiplayer testing much harder.
> 
> **Edit:** @Desour told me that this does not break local multiplayer, but I fail to see how I can ever expect two Minetest instances with the same cache directory to work together flawlessly if the databasse gets locked here.

Yep, it does not break anything.
The point here was that you can't run into serious synchronisation issues, ie. corrupted data or crashes.
The db is normally just locked for less than 100 ms, the blocked client will then just wait till the db is unlocked again.

**Comment by erlehmann** [comment]
I have written it in the issue, but if someone doubts that >1000 lines are not needed:

On Linux, `tmpreaper 30d cache` will remove files that have not been accessed in 30 days.

This will only work if the filesystem is not mounted `noatime`, but @Desour did indeed do that.

http://manpages.ubuntu.com/manpages/bionic/man8/tmpreaper.8.html

**Comment by erlehmann** [comment]
> The alternative would be to remove the old folder. It might make sense to do that.

If you import it into the database, maybe it does.

> I don't care about a one-time download.

I am a user who uses a slow metered mobile connection on their main machine.
Therefore I do care very much if Minetest wants to download over 500 MB again.

I have friends who had to wait several minutes to join servers. It is unfair to clear their cache!

> The db is normally just locked for less than 100 ms, the blocked client will then just wait till the db is unlocked again.

You wrote about getting kicked back to the main menu. So that does not happen?

**Comment by Desour** [comment]
> If you make 1000 players each wait additional 10 minutes to download all the cached stuff because the code expires their cache, then that is already a week of human lifetime you make them wait collectively, for no reason!

Well, I intend to live longer than 7 days. :P

> > The db is normally just locked for less than 100 ms, the blocked client will then just wait till the db is unlocked again.
> 
> You wrote about getting kicked back to the main menu. So that does not happen?

It does, but only if the db is locked long enough.

**Comment by erlehmann** [comment]
> Well, I intend to live longer than 7 days. :P

It is not about your life, it is affecting the lifetime of other people by making them wait collectively.

**Comment by Desour** [comment]
> > Well, I intend to live longer than 7 days. :P
> 
> It is not about your life, it is affecting the lifetime of other people by making them wait collectively.

I was referring to this exaggeration of yours:
> that is already a week of human lifetime you make them wait collectively

**Comment by hecktest** [comment]
I support this in principle. One thing that everybody is missing is that there is a hidden cost to generating lots of small files - it can trash the filesystem itself. An MT cache folder can contain tens of thousands of tiny files, each of them needing a whole entry in the FS. If you keep several MT folders side by side (PR tests, different versions, etc.) then each one of them will generate its own bag of files, even if you test offline since the client doesn't differentiate between single player and a server.

In a way, by not using a DB for the cache you are not avoiding the costs of maintaining one - you are merely offsetting them onto the FS, which is not equipped with domain specific knowledge that your program can use to optimize it.

Some filesystems are more affected by having tons of files in a single folder, others by just having a lot of small files in general. (like NTFS, which just grows its MFT and never shrinks it)

More often than not, the damage done to a filesystem in this way is nigh-irreversible short of just remaking the partition. So yes, for the sake of not damaging people's systems, we should be doing this. I personally banish each MT folder's cache to a partition created just for this sort of litter - but ideally, a program should not force anyone to do things like this to avoid the damage.

Assorted notes:
- Maximum cache size is a bad idea, because it introduces a (shared!) upper limit on a game's media size - there is no reason to assume any sort of upper bound, nothing is stopping a game from being several gigs in size. (other than people's willingness to wait for the download)
- Try to think about edge cases when designing the collector - big files, unique file sets, large game downloads, people reconnecting a lot, people coming back to a server after some time. A good solution might require a combination of heuristics that "score" certain files as less likely to get deleted. Look into generational garbage collectors as they deal with a similar problem. Consider holding on to entries for some time even after deleting the file, to detect cases where files re-appear frequently. (and bias them towards longer retention) Account for the entire download batch size when scoring a file, and how many servers it's been encountered on.
- Store the files in the DB, otherwise you're just gonna deal with the worst aspects of both approaches. You'll be paying for both the DB itself and the filesystem litter.

**Comment by erlehmann** [comment]
I just realized that there exists a program for storing files and metadata that creates “lots” of small files for its filesystem-backed cache. It works just fine, even on Windows: That program is called Git. You probably even use it yourself sometimes!

Git does not use a relational database like SQLite because that is the wrong tool for keeping track of blobs with a bit of metadata. Relational databases solve different problems than filesystems. Filesystems are very efficient key-value stores, so if you want content-addressable blobs, they are a reasonable choice. You might want to layer some abstraction on top of it, like making prefix subfolders or git packfiles – but it will probably not be as complex (i.e. featureful and possibly error-prone) as a full-blown relational database.

**Comment by TuringZauber** [comment]
(not a dev here) am I understanding correctly that this is for caching the mods which are downloaded from the server?
When I'm connecting to a server then it appears like Minetest is _always_ downloading _all_ the concent. So when I exit and rejoin again the join takes quite a while. I'm not sure what's the exact reason for this but I assume it is redownloading all the resource packs, mods and map?
I'd rather trade local storage for speed. Today consumer-grade terabytes are mere euros and large SSDs are becoming more and more widespread... Reading 500 MB from disk seems much better than downloading it from the internet...

**Comment by erlehmann** [comment]
@TuringZauber no idea why your rejoins take long, but minetest already has a filesystem-based cache.

This PR would replace that solution with a database and (IMO highly questionable) cache expiry logic.

I believe the original motivation for that approach is that the submitter has used the mount option `noatime`, which prevents the file system of saving when a file was accessed last. With the mount options `atime` or `relatime` (the latter seems to be the default on Debian GNU/Linux), deleting files that have not been accessed in a while is so simple that it is not even included as a feature in the game:

> On Linux, tmpreaper 30d cache will remove files that have not been accessed in 30 days.

**Comment by TuringZauber** [comment]
Thanks, @erlehmann. This explains a lot to me. Thanks for explaining about the cache rules. SQLite is faster than the file system and perhaps it's even a proper second level cache (i'm doubtful because client caches usually don't need SQL syntax support).

Idk why joining https://your-land.de/ is so slow. Perhaps they have speed problems unrelated to MT.

**Comment by erlehmann** [comment]
> SQLite is faster than the file system

I doubt that this is generally true and doubt it even matters here if it is true.

> i'm doubtful because client caches usually don't need SQL syntax support

Same. Even without atime support in the filesystem a cache eviction implementation can be *much* simpler.

> Idk why joining https://your-land.de/ is so slow. Perhaps they have speed problems unrelated to MT.

I suggest capturing the network traffic during a join with Wireshark.

Use the dissector `util/wireshark/minetest.lua` on the captured data.

**Comment by TuringZauber** [comment]
>> SQLite is faster than the file system
> I doubt that this is generally true

to quote what has already been stated in this thread: https://www.sqlite.org/fasterthanfs.html

> ... and doubt it even matters here if it is true.

it does!

> Use the dissector `util/wireshark/minetest.lua` on the captured data.

Cool, that's really useful!

**Comment by erlehmann** [comment]
@TuringZauber the article itself states:

> I/O performance varies widely depending on operating system and hardware. Make your own measurements before drawing conclusions.

I have written shell scripts myself that ran faster than some “SQLite is faster, I won't measure” software.

But I also have stored data in CSV and TSV files and got much better performance when switching to SQLite.

It always depends on the task at hand, i.e. any benchmark by the library authors can be true, yet useless.

**Comment by TuringZauber** [comment]
Cool, `random anecdotes unrelated to the issue at hand`! 
Thanks for sending me these I really needed these `random anecdotes unrelated to the issue at hand`.

**Comment by rubenwardy** [comment]
I suspect that we need more information to determine whether this is worth doing or not. Benchmarks for different situations would be helpful

**Comment by sfan5** [comment]
Closed for lack of benchmarks.
Coredev opinion is split on whether an sqlite db for cache is a good idea in the first place, alternative suggestions include only tracking metadata (for cache expiry) in sqlite.

**Comment by JeanPaulLucien** [comment]
Use database to store files inside is wrong and bad idea.

**Comment by fluxionary** [comment]
> Use database to store files inside is wrong and bad idea. That is not task for database.

@JeanPaulLucien the semantic difference between a file system and a database is not a large one, compared to the similarities. databases can be used as file systems, and vice versa, and the performance tradeoffs often lie in entirely different scopes. there's no objective reason to not e.g. store binary image data in a concrete database vs "files" that the operating system is aware of, or conversely to use a concrete database system like postgres to store image or model data vs. putting it into some sort of indexed file. filesystems have overheads, but databases also have overheads. without real performance testing on a wide variety of platforms, i don't think we should make too many assumptions about what an "optimal" architecture is.

**Comment by HybridDog** [comment]
There exists recent (2024-03-21 10:47:57) research about the performance of storing files in a database instead of file system: [Why Files If You Have a DBMS?](https://www.cs.cit.tum.de/fileadmin/w00cfj/dis/papers/blob.pdf)
They've benchmarked the differences between files stored in various databases and file systems, and they've invented something new for database systems which is faster than existing databases or file systems. I think it's implemented somewhere in [LeanStore](https://leanstore.io).
